### PR TITLE
#1169 make links a model

### LINKED ISSUE #1169: Schema - Make Link Model
### Description

We want to change the way we do links so that we can easily add and remove links to projects. 

### Acceptance Criteria

make Link Model with a name field as a string and a url field as a string and then remove all the individual links from the project model and replace it with a list of links.

### Proposed Solution

^

### Mocks

_No response_

### PR DESCRIPTION
**Pull request by Peyton-McKee**
## Changes

Edited the schema
added get all link types endpoint
added transformers
added seed data
changed transfomrers
added and adjusted query args
edited frontend view 
edited frontend edit view
refactored all of changes

## Notes

currently frontedn use field array isnt working, will it get it working hopefully before tomorrow morning

## Screenshots
Screen Shot 2023-06-04 at 3 10 01 PM

Screen Shot 2023-06-04 at 3 09 27 PM

Screen Shot 2023-06-04 at 3 10 33 PM

## To Do

Fix Bug with editing

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No linting errors / newline at end of file warnings
- [ ] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Screenshots of UI changes (see Screenshots section)
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] No `yarn.lock` changes (unless dependencies have changed)
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes #1169  (issue #)

### CODE REVIEW

File: src/backend/src/prisma/schema.prisma [hunk 1]
```diff
@@ -282,6 +281,28 @@ model Work_Package {
...
+
+model Link {
+  linkId      String   @id @default(uuid())
+  url         String
+  creatorId   Int
+  creator     User     @relation(name: "linkCreator", fields: [creatorId], references: [userId])
+  dateCreated DateTime @default(now())
+  linkTypeId  String
+  linkType    LinkType @relation(name: "linkTypes", fields: [linkTypeId], references: [linkTypeId])
+  projectId   Int
```
Comment (anthonybernardi) [inline src/backend/src/prisma/schema.prisma:302]: could be good for work packages too (imagine D&C with like github issue links or something idk)

File: src/backend/src/prisma/migrations/20230604173138_link_to_model/migration.sql [hunk 2]
```diff
@@ -0,0 +1,49 @@
+/*
```
Comment (anthonybernardi) [inline src/backend/src/prisma/migrations/20230604173138_link_to_model/migration.sql:1]: is the plan to just delete all the existing links?

File: src/backend/src/prisma/seed.ts [hunk 3]
```diff
@@ -160,10 +184,18 @@ const performSeed: () => Promise<void> = async () => {
     ['Decrease size by 90% from 247 cubic inches to 24.7 cubic inches'],
     ['Capable of absorbing 5000N in a head-on collision'],
     ['Cannot go further towards the rear of the car than the front roll hoop'],
-    'https://youtu.be/dQw4w9WgXcQ',
-    'https://youtu.be/dQw4w9WgXcQ',
-    'https://youtu.be/dQw4w9WgXcQ',
-    'https://youtu.be/dQw4w9WgXcQ',
+    [
+      {
+        linkId: '-1',
```
Comment (anthonybernardi) [inline src/backend/src/prisma/seed.ts:189]: kinda confusing if the id is supposed to be a string but you make it a number

File: src/backend/src/prisma/seed.ts [hunk 4]
```diff
@@ -182,10 +214,18 @@ const performSeed: () => Promise<void> = async () => {
     ['Decrease weight by 90% from 4.8 pounds to 0.48 pounds'],
     ['Provides removable section for easy access to the pedal box'],
     ['Compatible with a side-pod chassis design'],
-    'https://youtu.be/dQw4w9WgXcQ',
-    'https://youtu.be/dQw4w9WgXcQ',
-    'https://youtu.be/dQw4w9WgXcQ',
-    'https://youtu.be/dQw4w9WgXcQ',
+    [
+      {
+        linkId: '-1',
```
Comment (anthonybernardi) [inline src/backend/src/prisma/seed.ts:219]: same thing

File: src/backend/src/routes/projects.routes.ts [hunk 5]
```diff
@@ -35,10 +36,10 @@ projectRouter.post(
...
   body('otherConstraints.*.id').isInt({ min: -1 }).not().isString(),
   nonEmptyString(body('otherConstraints.*.detail')),
-  nonEmptyString(body('googleDriveFolderLink')),
-  nonEmptyString(body('slideDeckLink')),
-  nonEmptyString(body('bomLink')),
-  nonEmptyString(body('taskListLink')),
+  body('links').isArray(),
+  nonEmptyString(body('links.*.name')),
+  nonEmptyString(body('links.*.url')),
+  nonEmptyString(body('links.*.linkTypeId')),
```
Comment (anthonybernardi) [inline src/backend/src/routes/projects.routes.ts:42]: realistically we could probably have link type just be the name itself and make it @unique @id in the schema and that should work

no point in the link type id being a uuid which makes us add fields like this and stuff

File: src/backend/src/utils/auth.utils.ts [hunk 6]
```diff
@@ -78,3 +78,5 @@ export const getCurrentUser = async (res: Response): Promise<UserWithTeam> => {
   if (!user) throw new NotFoundException('User', userId);
   return user;
 };
+
+export type UserWithTeam = User & { teams: Team[] };
```
Comment (anthonybernardi) [inline src/backend/src/utils/auth.utils.ts:82]: hmm

File: src/frontend/src/components/LinkEditView.tsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  append: UseFieldArrayAppend<any, any>;
+  remove: UseFieldArrayRemove;
+}> = ({ name, ls, register, append, remove }) => {
+  const { isLoading, isError, error, data } = useAllLinkTypes();
+  if (isLoading || !data) return <LoadingIndicator />;
+  if (isError) return <ErrorPage message={error.message} />;
+
+  return (
+    <>
+      {ls.map((_element, i) => {
```
Comment (anthonybernardi) [inline src/frontend/src/components/LinkEditView.tsx:21]: not very pretty

**Review by anthonybernardi — COMMENTED** [review]
i like the concept of this

refactoring all the edit stuff in this same pr is pretty big tho, gotta make sure every single edge case still works right and we don't have good tests for them unfortunately

File: src/backend/src/prisma/migrations/20230604173138_link_to_model/migration.sql (hunk 2 above)
Comment (Peyton-McKee) [reply src/backend/src/prisma/migrations/20230604173138_link_to_model/migration.sql:1]: Negative i have to add the insertions manually once i finish everything else

File: src/backend/src/prisma/seed.ts (hunk 4 above)
Comment (Peyton-McKee) [reply src/backend/src/prisma/seed.ts:219]: This is just to symbolize a non created link

File: src/backend/src/routes/projects.routes.ts (hunk 5 above)
Comment (Peyton-McKee) [reply src/backend/src/routes/projects.routes.ts:42]: Thats true i didnt think about that

File: src/frontend/src/components/LinkEditView.tsx (hunk 7 above)
Comment (Peyton-McKee) [reply src/frontend/src/components/LinkEditView.tsx:21]: I copied and pasted this from the react hook edit list component we have, we should prob refactor but im not very familiar with use field array yet

File: src/backend/src/utils/auth.utils.ts (hunk 6 above)
Comment (Peyton-McKee) [reply src/backend/src/utils/auth.utils.ts:82]: this is from a different PR i just moved it to this file

File: src/backend/src/utils/changes.utils.ts [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+              seenOld.get(change.changeListValue.comparator) || 'null',
+              seenNew.get(change.changeListValue.comparator) || 'null'
+            )
+          : `${change.type} ${nameOfField} "${change.changeListValue.displayValue}"`;
+      return { changeRequestId: crId, implementerId, wbsElementId, detail };
+    })
+  };
+};
+
+// // returns a change if the old and new dates are different, otherwise return undefined
```
Comment (RChandler234) [inline src/backend/src/utils/changes.utils.ts:140]: what's up with all the commented out code?

**Review by RChandler234 — CHANGES_REQUESTED** [review]
All the code looks good (I can do a closer look later). Functionality-wise, I was thinking that a project should at least have a list of required linkTypes that could be edited via admin tools or smthn. The fully customizable links looks cool, but I think it might be more useful for the create project form to just be a template they need to fill out, that way the projects are more standardized and leads don't even have to think about what links they want to add. I think the fully customizable links where they act like description bullets might be more useful on smthn like a work package like Anthony was talking about where the details will be more specific to the type of work that's being done.

tldr: I like the flexibility this gives us in the backend/ all the changes abstraction is sweet, but I think we should give some more thought to how we wanna expose this new functionality to the users

File: src/backend/src/services/projects.services.ts [hunk 9]
```diff
@@ -341,15 +329,57 @@ export default class ProjectsService {
...
+      updatedProject.projectId,
+      'projectIdOtherConstraints'
+    );
+    await editDescriptionBullets(
+      goalsChangeJson.editedElements
+        .concat(featuresChangeJson.editedElements)
+        .concat(otherConstraintsChangeJson.editedElements)
     );
 
+    await linkChanges.addedElements.forEach(async (link) => {
```
Comment (Peyton-McKee) [inline src/backend/src/services/projects.services.ts:353]: do we want to make these three things its own function

File: src/backend/src/transformers/links.transformer.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
+import { Prisma } from '@prisma/client';
```
Comment (Peyton-McKee) [inline src/backend/src/transformers/links.transformer.ts:1]: also might want to just put these two transformers in the same file

File: src/frontend/src/components/LinkEditView.tsx [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        return (
+          <Grid item sx={{ display: 'flex', alignItems: 'center', mb: '5px' }}>
+            <TextField
```
Comment (Peyton-McKee) [inline src/frontend/src/components/LinkEditView.tsx:37]: @RChandler234 havnt implemented the rigid structure for editing the project with the links, specifically what are you thinking of.

Comment (Peyton-McKee) [reply src/frontend/src/components/LinkEditView.tsx:37]: do you think it shouldnt be a use field array

Comment (RChandler234) [reply src/frontend/src/components/LinkEditView.tsx:35]: I'm fine with the use field array stuff, but there should be some links that are required on all projects (and we should have a corresponding tool in admin tools for admins to set this). I like how it looks in your screenshots, but we would just disable the ability to edit the type or delete for the required links (grey out the select with it set to the proper type, and remove the delete button)

Comment (RChandler234) [reply src/frontend/src/components/LinkEditView.tsx:35]: and then they would be able to add as many extra links as they want if they feel it's useful

File: src/backend/src/services/projects.services.ts (hunk 9 above)
Comment (RChandler234) [reply src/backend/src/services/projects.services.ts:353]: yeah that would make it a little cleaner

File: src/shared/src/types/project-types.ts [hunk 12]
```diff
@@ -75,3 +72,30 @@ export interface DescriptionBullet {
...
+  url: string;
+}
+
+export interface LinkCreateArgs {
+  linkId: string;
+  linkTypeName: string;
+  url: string;
+}
+
+export enum RequiredLinkType {
```
Comment (RChandler234) [inline src/shared/src/types/project-types.ts:97]: required should just be a boolean that can be toggled on a linkType, having it as an enum here is weird

File: src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditContainer.tsx [hunk 13]
```diff
@@ -84,15 +101,18 @@ const ProjectEditContainer: React.FC<ProjectEditContainerProps> = ({ project, ex
...
-      slideDeckLink,
-      bomLink,
-      taskListLink,
-      googleDriveFolderLink: gDriveLink,
       summary,
       crId: query.get('crId') || '',
       rules: project.rules.map((rule) => {
         return { rule };
       }),
+      links: projectLinks.map((link) => {
```
Comment (RChandler234) [inline src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditContainer.tsx:109]: also this map is unnecessary, you're just making a new object that's the same as the old one then returning it

File: src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditContainer.tsx [hunk 14]
```diff
@@ -73,7 +73,24 @@ const ProjectEditContainer: React.FC<ProjectEditContainerProps> = ({ project, ex
...
+    return {
+      linkId: link.linkId,
+      url: link.url,
+      linkTypeName: link.linkType.name
+    };
+  });
+
+  const linkTypes = projectLinks.map((link) => link.linkTypeName);
+
+  Object.values(RequiredLinkType).forEach((linkType) => {
```
Comment (RChandler234) [inline src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditContainer.tsx:88]: is this adding the required link type if it's not there? kinda weird, I'd find a better way to enfore that the required types are included (maybe in yup) or at least put a comment saying what you're doing here

File: src/frontend/src/components/LinkEditView.tsx [hunk 15]
```diff
@@ -0,0 +1,70 @@
+import { useAllLinkTypes } from '../hooks/projects.hooks';
+import LoadingIndicator from './LoadingIndicator';
+import ErrorPage from '../pages/ErrorPage';
+import { Button, Grid, IconButton, MenuItem, TextField } from '@mui/material';
+import { FieldArrayWithId, UseFieldArrayAppend, UseFieldArrayRemove, UseFormRegister } from 'react-hook-form';
+import DeleteIcon from '@mui/icons-material/Delete';
+import { LinkCreateArgs } from 'shared';
+
+const LinkEditView: React.FC<{
```
Comment (RChandler234) [inline src/frontend/src/components/LinkEditView.tsx:9]: LinksEditView

File: src/frontend/src/components/LinkEditView.tsx [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+  remove: UseFieldArrayRemove;
+  links: LinkCreateArgs[];
+}> = ({ ls, register, append, remove, links }) => {
+  const { isLoading, isError, error, data } = useAllLinkTypes();
+  if (isLoading || !data) return <LoadingIndicator />;
+  if (isError) return <ErrorPage message={error.message} />;
+
+  return (
+    <>
+      {ls.map((_element, i) => {
```
Comment (RChandler234) [inline src/frontend/src/components/LinkEditView.tsx:32]: checking whether something is required by checking the index is WAY too jank. I think adding the required field on linkType would make this easier, you could just check that to see what should be enabled/disabled

File: src/frontend/src/components/LinkEditView.tsx [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+        return (
+          <Grid item sx={{ display: 'flex', alignItems: 'center', mb: '5px' }}>
+            <TextField
+              {...register(`links.${i}.linkTypeName`, { required: true })}
+              sx={{ minWidth: '200px', mr: '5px' }}
+              select
+              defaultValue={links[i].linkTypeName}
+              disabled={i < 3}
+            >
```
Comment (RChandler234) [inline src/frontend/src/components/LinkEditView.tsx:41]: why is this a textfield instead of a select?

File: src/frontend/src/components/LinkEditView.tsx (hunk 15 above)
Comment (RChandler234) [reply src/frontend/src/components/LinkEditView.tsx:9]: also general note on this component, it's kinda weird that you can add multiple of a link type, cause then it's just two links with the same label like this. I think we should make it only one of each link type per project
image

**Review by RChandler234 — CHANGES_REQUESTED** [review]
just reviewed the links stuff, haven't even touched changes. If you could give me a quick rundown of how you refactored it, it'd make reviewing a lot easier. For future reference two things this separate and complicated should NEVER be in the same PR

**Comment by RChandler234** [comment]
also prettier is failing

File: src/backend/src/prisma/schema.prisma [hunk 18]
```diff
@@ -288,6 +287,27 @@ model Work_Package {
...
+model Link {
+  linkId       String   @id @default(uuid())
+  url          String
+  creatorId    Int
+  creator      User     @relation(name: "linkCreator", fields: [creatorId], references: [userId])
+  dateCreated  DateTime @default(now())
+  linkTypeName String
+  linkType     LinkType @relation(name: "linkTypes", fields: [linkTypeName], references: [name])
+  projectId    Int
+  project      Project  @relation(name: "links", fields: [projectId], references: [projectId])
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply src/backend/src/prisma/schema.prisma:308]: idk we only have links on projects rn.

File: src/frontend/index.html [hunk 19]
```diff
@@ -10,6 +10,7 @@
     <link rel="manifest" href="/manifest.json" />
     <link href="https://fonts.googleapis.com/css?family=Lato" rel="stylesheet" />
     <link href="https://fonts.googleapis.com/css?family=Oswald" rel="stylesheet" />
+    <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply src/frontend/index.html:13]: allowing icons by name

File: src/backend/src/prisma/schema.prisma [hunk 20]
```diff
@@ -288,6 +287,27 @@ model Work_Package {
   stage              Work_Package_Stage?
 }
 
+model LinkType {
+  name        String   @id @unique
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply src/backend/src/prisma/schema.prisma:291]: anthony suggested this

File: src/backend/src/utils/auth.utils.ts [hunk 21]
```diff
@@ -99,3 +99,5 @@ export const getCurrentUserWithUserSettings = async (res: Response): Promise<Use
...
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply src/backend/src/utils/auth.utils.ts:103]: I thought i removed this in the merge conflicts whoops

File: src/shared/src/types/project-types.ts (hunk 12 above)
Comment (Peyton-McKee) [reply src/shared/src/types/project-types.ts:97]: mm thats smart actually

File: src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditContainer.tsx (hunk 13 above)
Comment (Peyton-McKee) [reply src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditContainer.tsx:109]: no this is required for useFieldArray

File: src/frontend/src/components/LinkEditView.tsx (hunk 15 above)
Comment (Peyton-McKee) [reply src/frontend/src/components/LinkEditView.tsx:9]: hmmm maybe

File: src/frontend/src/components/LinkEditView.tsx (hunk 17 above)
Comment (Peyton-McKee) [reply src/frontend/src/components/LinkEditView.tsx:41]: copied james code for editable list component

File: src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditContainer.tsx (hunk 14 above)
Comment (Peyton-McKee) [reply src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditContainer.tsx:88]: yes that is what this is doing. its just supplying default values, not necessarily enforcing them

File: src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditContainer.tsx (hunk 13 above)
Comment (Peyton-McKee) [reply src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditContainer.tsx:109]: I think....

Comment (RChandler234) [reply src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditContainer.tsx:115]: what is it doing then?

File: src/frontend/src/components/LinkEditView.tsx (hunk 15 above)
Comment (RChandler234) [reply src/frontend/src/components/LinkEditView.tsx]: we gotta fix this problem where u can have two links with the same name (it's confusing), or remove the ability to have more than one link type per project

File: src/backend/src/prisma/schema.prisma (hunk 18 above)
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/backend/src/prisma/schema.prisma:309]: yeah, but what if we wanna have it on other stuff in the future? seems like we should make it as generic as possible while we're updating it, maybe kind of like how it's done on description bullets, we could have projects and wps on a link and have them both be optional

File: src/frontend/src/components/LinksEditView.tsx [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+      requiredLinkTypeNames.includes(linkTypeName) &&
+      !currentLinkTypeNames.includes(linkTypeName, currentLinkTypeNames.indexOf(linkTypeName) + 1)
+    );
+  };
+
+  return (
+    <>
+      {ls.map((_element, i) => {
+        return (
+          <Grid item sx={{ display: 'flex', alignItems: 'center', mb: '5px' }}>
```
Comment (RChandler234) [inline src/frontend/src/components/LinksEditView.tsx:40]: this shouldn't be a grid item because it's not in a grid, a box would make more sense. Also each item in this map needs to have a unique key

File: src/frontend/src/components/LinksEditView.tsx [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+          </Grid>
+        );
+      })}
+      <Button
+        variant="contained"
+        color="success"
+        onClick={() => append({ linkId: '-1', url: '', linkTypeName: '-1' })}
+        sx={{ my: 2, width: 'max-content' }}
+      >
+        + Add New Bullet
```
Comment (RChandler234) [inline src/frontend/src/components/LinksEditView.tsx:75]: Add new link

File: src/frontend/src/components/LinksEditView.tsx [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+              disabled={isRequired(i)}
+            >
+              {linkTypes.map((linkType) => (
+                <MenuItem key={linkType.name} value={linkType.name}>
+                  {linkType.name}
+                </MenuItem>
+              ))}
+            </Select>
+            <TextField required fullWidth autoComplete="off" {...register(`links.${i}.url`, { required: true })} />
+            {!isRequired(i) && (
```
Comment (RChandler234) [inline src/frontend/src/components/LinksEditView.tsx:54]: if it's not required, the textfield should just cutoff where it does with the delete button there so all the textfields sizes line up. Like this: 
image

File: src/frontend/src/components/LinksEditView.tsx [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+  if (isLoading || !linkTypes) return <LoadingIndicator />;
+  if (isError) return <ErrorPage message={error.message} />;
+
+  const requiredLinkTypeNames = getRequiredLinkTypeNames(linkTypes);
+
+  const links = watch('links');
+
+  const currentLinkTypeNames = links.map((link) => link.linkTypeName);
+
+  const isRequired = (index: number) => {
```
Comment (RChandler234) [inline src/frontend/src/components/LinksEditView.tsx:27]: add a comment specifying that this checks whether a link has a required type and has not already appeared in the list of links once

**Review by RChandler234 — CHANGES_REQUESTED** [review]

File: src/backend/src/prisma/schema.prisma (hunk 18 above)
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply src/backend/src/prisma/schema.prisma:309]: yeah i can do that

File: src/frontend/src/components/LinksEditView.tsx [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+  register: UseFormRegister<ProjectEditFormInput>;
+  watch: UseFormWatch<ProjectEditFormInput>;
+  append: UseFieldArrayAppend<any, any>;
+  remove: UseFieldArrayRemove;
+}> = ({ ls, register, append, remove, watch }) => {
+  const { isLoading, isError, error, data: linkTypes } = useAllLinkTypes();
+  if (isLoading || !linkTypes) return <LoadingIndicator />;
+  if (isError) return <ErrorPage message={error.message} />;
+
+  const requiredLinkTypeNames = getRequiredLinkTypeNames(linkTypes);
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply src/frontend/src/components/LinksEditView.tsx:21]: no because i need all the links not just the required ones i was thinking about this

File: src/frontend/src/components/LinksEditView.tsx [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+                  {linkType.name}
+                </MenuItem>
+              ))}
+            </Select>
+            <TextField required fullWidth autoComplete="off" {...register(`links.${i}.url`, { required: true })} />
+            {!isRequired(i) && (
+              <IconButton
+                type="button"
+                onClick={() => {
+                  console.log(i);
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply src/frontend/src/components/LinksEditView.tsx:58]: still debugging but yes

File: src/frontend/src/components/LinksEditView.tsx (hunk 24 above)
Comment (Peyton-McKee) [reply src/frontend/src/components/LinksEditView.tsx:54]: ok

File: src/frontend/src/components/LinkEditView.tsx (hunk 15 above)
Comment (Peyton-McKee) [reply src/frontend/src/components/LinkEditView.tsx]: i think wed have to only allow one type per element because idk how wed change the name

File: src/frontend/src/components/LinksEditView.tsx (hunk 24 above)
Comment (Peyton-McKee) [reply src/frontend/src/components/LinksEditView.tsx:54]: how tf did u do that

File: src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditDetails.tsx [hunk 28]
```diff
@@ -36,7 +35,7 @@ const ProjectEditDetails: React.FC<ProjectEditDetailsProps> = ({
     <PageBlock title="Project Details">
       <Grid container xs={12} sx={{ my: 1 }}>
         <Grid item xs={8}>
-          <FormControl sx={{ width: '90%' }}>
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditDetails.tsx:39]: whoops didnt mean to do that mustve been a merge conflict

File: src/backend/src/services/work-packages.services.ts [hunk 29]
```diff
@@ -464,33 +453,35 @@ export default class WorkPackagesService {
...
+        where: { descriptionId: { in: deletedElements.map((descriptionBullet) => descriptionBullet.id) } },
         data: { dateDeleted: new Date() }
       });
     }
 
-    addDescriptionBullets(
-      expectedActivitiesChangeJson.addedDetails,
+    await addDescriptionBullets(
+      expectedActivitiesChangeJson.addedElements.map((descriptionBullet) => descriptionBullet.detail),
       updatedWorkPackage.workPackageId,
```
Comment (RChandler234) [inline src/backend/src/services/work-packages.services.ts:474]: how can we pass in a workPackageId here but a projectId in the project services file?

File: src/backend/src/services/work-packages.services.ts [hunk 30]
```diff
@@ -464,33 +453,35 @@ export default class WorkPackagesService {
...
+    await addDescriptionBullets(
+      deliverablesChangeJson.addedElements.map((descriptionBullet) => descriptionBullet.detail),
       updatedWorkPackage.workPackageId,
       'workPackageIdDeliverables'
     );
-    editDescriptionBullets(
-      expectedActivitiesChangeJson.editedIdsAndDetails.concat(deliverablesChangeJson.editedIdsAndDetails)
-    );
+
+    await editDescriptionBullets(expectedActivitiesChangeJson.editedElements.concat(deliverablesChangeJson.editedElements));
```
Comment (RChandler234) [inline src/backend/src/services/work-packages.services.ts:484]: add comments above add and edit like I suggested in projects for better readabiltiy

**Review by RChandler234 — CHANGES_REQUESTED** [review]
damn that changes shit is dense, good job refactoring all that, it's a lot more doable to follow

File: src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditDetails.tsx [hunk 31]
```diff
@@ -47,84 +46,40 @@ const ProjectEditDetails: React.FC<ProjectEditDetailsProps> = ({
...
         <Grid item xs={4}>
-          <FormControl sx={{ width: '90%' }}>
+          <FormControl fullWidth>
             <FormLabel>Budget</FormLabel>
             <ReactHookTextField
               name="budget"
               control={control}
               type="number"
               placeholder="Enter budget..."
-              icon={<AttachMoneyIcon />}
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply src/frontend/src/pages/ProjectDetailPage/ProjectEdit/ProjectEditDetails.tsx:57]: oh i was getting an error because of the merge conflict with megans PR

**Review by RChandler234 — CHANGES_REQUESTED** [review]
I'm getting an error with the seed file when I run yarn prisma:reset

**Review by RChandler234 — CHANGES_REQUESTED** [review]
not sure if this is smthn you did, but can you fix it while you're in here:
image

I tested all the functionality, everything working great, so after this small thing is fixed, I'll go ahead and merge. Can you also make a ticket to add an admin tool to add link types/ required link types

File: src/backend/src/utils/changes.utils.ts [hunk 32]
```diff
@@ -0,0 +1,129 @@
...
+  );
+  const seenNew = new Map<string, string>(
+    newArray.map((changeListValue) => [changeListValue.comparator, changeListValue.displayValue])
+  );
+
+  const changes: { changeListValue: ChangeListValue<T>; type: ChangeType }[] = [];
+
+  oldArray.forEach((changeListValue) => {
+    if (!seenNew.has(changeListValue.comparator)) {
+      console.log('removed ', changeListValue);
```
Comment (RChandler234) [inline src/backend/src/utils/changes.utils.ts:94]: oop and remove this

**Review by RChandler234 — CHANGES_REQUESTED** [review]

**Review by RChandler234 — APPROVED** [review]
